pkg/compactv2/compactor.go

```diff
@@ -0,0 +1,418 @@
+func (e errChunk) Appender() (chunkenc.Appender, error)         { return nil, e.err.err }
+func (e errChunk) Iterator(chunkenc.Iterator) chunkenc.Iterator { return e.err }
+func (e errChunk) NumSamples() int                              { return 0 }
+func (e errChunk) Compact()                                     {}
+
+func (l *lazyPopulatableChunk) populate() {
+	// TODO(bwplotka): In most cases we don't need to parse anything, just copy. Extend reader/writer for this.
+	var err error
+	l.populated, err = l.cr.Chunk(l.m.Ref)
+	if err != nil {
+		l.m.Chunk = errChunk{err: errChunkIterator{err: errors.Wrapf(err, "cannot populate chunk %d", l.m.Ref)}}
+		return
+	}
+
+	l.m.Chunk = l.populated
+}
+
+func (l *lazyPopulatableChunk) Bytes() []byte {
+	if l.populated == nil {
+		l.populate()
+	}
+	return l.populated.Bytes()
+}
+
+func (l *lazyPopulatableChunk) Encoding() chunkenc.Encoding {
+	if l.populated == nil {
+		l.populate()
+	}
+	return l.populated.Encoding()
+}
+
+func (l *lazyPopulatableChunk) Appender() (chunkenc.Appender, error) {
+	if l.populated == nil {
+		l.populate()
+	}
+	return l.populated.Appender()
+}
+
+func (l *lazyPopulatableChunk) Iterator(iterator chunkenc.Iterator) chunkenc.Iterator {
+	if l.populated == nil {
+		l.populate()
+	}
+	return l.populated.Iterator(iterator)
+}
+
+func (l *lazyPopulatableChunk) NumSamples() int {
+	if l.populated == nil {
+		l.populate()
+	}
+	return l.populated.NumSamples()
+}
+
+func (l *lazyPopulatableChunk) Compact() {
+	if l.populated == nil {
+		l.populate()
+	}
+	l.populated.Compact()
+}
+
+func (w *Compactor) write(ctx context.Context, symbols index.StringIter, populatedSet storage.ChunkSeriesSet, sWriter block.SeriesWriter, p ProgressLogger) error {
+	var (
+		chks []chunks.Meta
+		ref  uint64
+	)
+
+	for symbols.Next() {
+		if err := sWriter.AddSymbol(symbols.At()); err != nil {
+			return errors.Wrap(err, "add symbol")
+		}
+	}
+	if err := symbols.Err(); err != nil {
+		return errors.Wrap(err, "symbols")
+	}
+
+	// Iterate over all sorted chunk series.
+	for populatedSet.Next() {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		s := populatedSet.At()
+		chksIter := s.Iterator()
+		chks = chks[:0]
+		for chksIter.Next() {
+			// We are not iterating in streaming way over chunk as it's more efficient to do bulk write for index and
+			// chunk file purposes.
+			chks = append(chks, chksIter.At())
+		}
+
+		if chksIter.Err() != nil {
+			return errors.Wrap(chksIter.Err(), "chunk iter")
+		}
+
+		// Skip the series with all deleted chunks.
+		if len(chks) == 0 {
+			// All series will be ignored.
+			p.SeriesProcessed()
+			continue
+		}
+
+		if err := sWriter.WriteChunks(chks...); err != nil {
+			return errors.Wrap(err, "write chunks")
+		}
+		if err := sWriter.AddSeries(ref, s.Labels(), chks...); err != nil {
+			return errors.Wrap(err, "add series")
+		}
+		for _, chk := range chks {
+			if err := w.chunkPool.Put(chk.Chunk); err != nil {
```

Comment:
We put chunks to the pool but I don't find we call `get` to this pool

Reply:
It's in `tsdb.Open` (: